packages/xstate-react/src/useMachine.ts

```diff
@@ -0,0 +1,246 @@
+import { useState, useEffect, useRef, useLayoutEffect } from 'react';
+import {
+  interpret,
+  EventObject,
+  StateMachine,
+  State,
+  Interpreter,
+  InterpreterOptions,
+  MachineOptions,
+  StateConfig,
+  Typestate,
+  ActionObject,
+  ActionFunction,
+  ActionMeta
+} from 'xstate';
+import useConstant from './useConstant';
+import { partition } from 'xstate/lib/utils';
+
+enum ReactEffectType {
+  Effect = 1,
+  LayoutEffect = 2
+}
+
+export interface ReactActionFunction<TContext, TEvent extends EventObject> {
+  (
+    context: TContext,
+    event: TEvent,
+    meta: ActionMeta<TContext, TEvent>
+  ): () => void;
+  __effect: ReactEffectType;
+}
+
+export interface ReactActionObject<TContext, TEvent extends EventObject>
+  extends ActionObject<TContext, TEvent> {
+  exec: ReactActionFunction<TContext, TEvent>;
+}
+
+export function asEffect<TContext, TEvent extends EventObject>(
+  exec: ActionFunction<TContext, TEvent>
+): ReactActionFunction<TContext, TEvent> {
+  const effectExec: unknown = (context, event, meta) => {
+    // don't execute; just return
+    return () => {
+      return exec(context, event, meta);
+    };
+  };
+
+  Object.defineProperties(effectExec, {
+    name: { value: `effect:${exec.name}` },
+    __effect: { value: ReactEffectType.Effect }
+  });
+
+  return effectExec as ReactActionFunction<TContext, TEvent>;
+}
+
+export function asLayoutEffect<TContext, TEvent extends EventObject>(
+  exec: ActionFunction<TContext, TEvent>
+): ReactActionFunction<TContext, TEvent> {
+  const effectExec: unknown = (context, event, meta) => {
+    // don't execute; just return
+    return () => {
+      return exec(context, event, meta);
+    };
+  };
+
+  Object.defineProperties(effectExec, {
+    name: { value: `layoutEffect:${exec.name}` },
+    __effect: { value: ReactEffectType.LayoutEffect }
+  });
+
+  return effectExec as ReactActionFunction<TContext, TEvent>;
+}
+
+export type ActionStateTuple<TContext, TEvent extends EventObject> = [
+  ReactActionObject<TContext, TEvent>,
+  State<TContext, TEvent>
+];
+
+function executeEffect<TContext, TEvent extends EventObject>(
+  action: ReactActionObject<TContext, TEvent>,
+  state: State<TContext, TEvent>
+): void {
+  const { exec } = action;
+  const originalExec = exec!(state.context, state._event.data, {
```

Comment:
I'm worried that those currently are called with "settled" arguments whereas regular actions are called with the current arguments from the time when a transition was selected. This makes those different from other builtin actions.

It would be IMHO preferable to have those behaving more like the rest of builtin actions. A really quick idea of how we could solve this:

Introduce a new `capture` action (tentative name). It could then be "resolved" by `resolveActions` and actually executed by `exec` in the Interpreter class. In the future, we could consider "capturing" all custom actions to make them behave closer to the builtin ones in this regard. I think this is quite important because it seems quite expected to have FOO `event`'s type here at all times:
```
on: { FOO: { actions: (ctx, event) => {} } }
```

Reply:
To clarify w/r/t determinism: an action executed immediately should behave exactly the same as that same action executed e.g., 20ms later. This is even true in React, in which effects might be executed with stale data still.

Reply:
Hm, not sure if we are on the same page here, so maybe let me rephrase my concerns with an example:
```js
FOO: {
  actions: [
    assign((ctx1, ev1) => {},
    (ctx2, ev2) => {}),
    raise(/**/)
  ]
}
```
Given this example `ev1 !== ev2` because the second action, the one which is not a builtin one, gets executed (and arguments are provided to it) when the machine "settles", whereas for builtin actions we "capture" current (from the moment when a transition has happened) context/event. In a similar fashion the problem doesn't only impact event argument, but also context - for `assign` (in v5) this would refer to a context available at the time of selecting a transition whereas for non-core action this receives computed context after things settle (so if for example `assign` follows non-core action then that non-core action receives "updated" context, but imho it shouldn't).

IMHO the difference described above leads to non-determinism and makes things harder to reason about.

Reply:
I see what you mean now. Let's see if I can change it.

Reply:
> custom actions are not called with intermediate context, although they should be. (Built-in actions like send(), log(), choose(), etc. are, though).

Right, but as those new actions here are sort-of builtins I think they should share the semantics with builtin actions, but right now they share it with custom ones.

> This needs to be fixed in xstate (and is a breaking change, so V5), not in here.

Agreed as to the custom ones.

>  I think you're confusing how the actions get resolved with how the actions get executed.

Yes, definitely part of my confusion. But if we in v5 plan to fix this also for custom actions then there won't be much a difference between what arguments are given to both builtin and custom actions as the resolving step would capture arguments and executing would just call what with what got captured - there wouldn't be any different set of arguments given when executing.